MSTransferor: support input data placement with Rucio

amaltaro opened the pull request:
Fixes #9725 
Fixes #9461

#### Status
In development

#### Description
Summary of changes are:
* implemented a Rucio wrapper based on pycurl for concurrent requests; only a few very needed APIs implemented so far;
* when listing Rucio replication rules, it could be that rules in state STUCK and SUSPENDED will not be considered and a new rule can be created;
* updated MSCore `__init__` method to support sub-classes to provide service names that they do not want to be initialized;
* updated MSCore to initialize a Rucio object;
* support RSE quotas update based on Rucio (using quota limit and usage)
* added all the necessary logic in MSTransferor to deal with either PhEDEx-based or Rucio-based input data placement;
  * including check on the RSE quota and usage
  * relying on a new Rucio account: `wmcore_transferor`
  * logic for the pileup is: if the container is not locked anywhere within the SiteWhitelist (with that Rucio acct), then we make a new rule with grouping=ALL against one specific RSE
  * logic for the primary is: find all the blocks and make a rule for all of them against all RSEs to be used (grouping=DATASET)
  * logic for the primary + parent is: create chunks of primary + parent and make a rule against one specific RSE (other chunks go to other RSEs), using grouping=DATASET
* Rucio rules are created in asynchronous mode, `ask_approval` always False;
* Rucio rules - in non production server/mode - will contain a lifetime of 24h;
* Rucio RSE quota and rucio rules are created with the python client APIs; container size, list of blocks and their sizes and current rules locking data is done via RESTful APIs;
* renew the Rucio token if it's to expiry within 30min;
* Todor/Kenyi/I are notified on large data placement;
* on the Rucio wrapper, support `lifetime` and `asynchronous` parameters for replication rule creation;

In addition to that, there is:
* fixed a bug in MSMonitor when counting workflows that had to be skipped;
* fixed a bug, inconsistent progress between Rucio and PhEDEx transfers. Now their progress is expressed in percentage, thus between 0 and 100;
* added a debug line to print some extra information about rules being evaluated

It depends on the following new configuration parameters:
* `useRucio`: to enable/disable Rucio as a DM service (if disabled, PhEDEx is used)
* `rulesLifetime`: defines an expiration time for all rules created by MSTransferor. Production rules (against the production Rucio server) have no expiration time.

#### Is it backward compatible (if not, which system it affects?)
yes

#### Related PRs
none

#### External dependencies / deployment changes
Deployment changes: 
https://github.com/dmwm/deployment/pull/924
https://github.com/dmwm/deployment/pull/942

amaltaro committed:
MSTransferor: support input data placement with Rucio

Implement whole logic of input data placement with Rucio

fix import from PycurlRucio

clean init file

minor aesthetic changes

import absolute_import

one more absolute_import

copy cert/ckey methods over to pycurlRucio

useRucio flag might not be in the configuration file

apply Todors suggestions

use getattr for self.phedex

set self.phedex instead

Rucio initialization message

complete RSE limits/usage logic

fix key name name -> rse

bytes_remaining is always 0, set it to quota

fix getRucioToken

getdata has no decode parameter

fix getPileupDatasetSizesRucio function

fix getDatasetBlocksRucio and getPileupSubscriptionsRucio

fix getBlockReplicasAndSizeRucio method

New pycurl function listReplicationRules

fix listReplicationRule fail case

API to fetch blocks and their sizes given a container

Make the correct calls for PhEDEx and Rucio

integrate pickRSE

update long=1 to long=True

fix logic for renewing the rucio token

RequestInfo child class of MSCore; skip some service initialization

fix keyword arguments

fix MSMonitor NoneType; add debug rule log

bugfixes to rule monitoring; rule creation; others

support lifetime parameter for rule replicas

fix MSOutput initialization

more fixes to the pycurl rucio function; print grouping of the rules

Make Rucio rule completion consistent with PhEDEx, percentage

stuck_at returns a string datetime format

minor logging fix

amaltaro commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -415,45 +391,143 @@ def makeTransferRequest(self, wflow):
         return success, response
 
-    def _subscribeData(self, subscriptionObj, wflowName, dsetName):
+    def makeTransferRucio(self, wflow, dataIn, subLevel, blocks, dataSize, nodes, nodeIdx):
```
@todor-ivanov FYI Todor, this is more or less what you will need for the output data placement

todor-ivanov replied:
Thanks @amaltaro  I am looking at your code now.

amaltaro commented:
For some reason, the `import cert, ckey` from PycurlRucio does not want to work, I get the same import error in my container too:
```
Traceback (most recent call last):
  File "/home/dmwm/unittestdeploy/wmagent/1.3.6.pre4/sw/slc7_amd64_gcc630/external/py2-nose/1.3.7-comp3/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/dmwm/unittestdeploy/wmagent/1.3.6.pre4/sw/slc7_amd64_gcc630/external/py2-nose/1.3.7-comp3/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/dmwm/unittestdeploy/wmagent/1.3.6.pre4/sw/slc7_amd64_gcc630/external/py2-nose/1.3.7-comp3/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/MicroService_t/DataStructs_t/Workflow_t.py", line 8, in <module>
    from WMCore.MicroService.DataStructs.Workflow import Workflow
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/MicroService/DataStructs/Workflow.py", line 10, in <module>
    from WMCore.MicroService.Unified.Common import getMSLogger, gigaBytes
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/MicroService/Unified/Common.py", line 18, in <module>
    from WMCore.MicroService.Tools.PycurlRucio import getPileupDatasetSizesRucio, getPileupSubscriptionsRucio, \
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/MicroService/Tools/PycurlRucio.py", line 20, in <module>
    from WMCore.MicroService.Unified.Common import cert, ckey
ImportError: cannot import name cert
```

it must be missing some subtle detail of the import behavior.
As a workaround, I copied those 2 methods over to the PycurlRucio module.

@todor-ivanov I believe there will be further changes to this PR. But just in case you can have a look at these changes by tomorrow morning, it would be great! Thanks

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -680,45 +650,143 @@ def makeTransferRequest(self, wflow):
+            res = self.rucio.createReplicationRule(dids, rseExpr, **rule)
+            if res:
+                self.logger.info("Rule replication created under id: %s", res[0])
+                transferId = res[0]
```
Now this one I consider it dangerous. Taking only the first ruleId. One situation where it won't work for sure (especially after the changes I made here [1]) is, if we face a `DuplicateRule` exception internally in the `createReplicationRule` method. Then it for sure will return a list of `ruleIds`, but they will have also historical meaning and one needs to make additional parsing to check which is the one that he actually needs.

[1]
https://github.com/dmwm/WMCore/pull/9772/commits/6bfc5cbb6e797a27ad441a6556fdf61b5fd4db4a#diff-9ac27fcf16ebc6591102e67cfa2e1212R517

amaltaro replied:
That's a very good point. I'm updating the code.

todor-ivanov commented on src/python/WMCore/MicroService/Tools/PycurlRucio.py:
```diff
@@ -0,0 +1,267 @@
+            continue
+        if locationByDset.get(dataset, "non-None") is None:
+            # then one of the block requests failed, skip the whole dataset
+            continue
```
Shouldn't we log our actions here too before we just continue?

amaltaro replied:
It will get logged at the `RequestInfo` level, but I agree that it's a good idea to log which query failed here. Log level a few lines above needs to be changed to error as well.

amaltaro replied:
Actually, this `if` will only be executed if the previous one has been evaluated to true (row[data] is None), for the same dataset. This means, one of the block level calls failed for this dataset, and in that case, we simply have to mark that dataset as "unresolved" for the moment and retry in the next cycle.

todor-ivanov reviewed (changes requested):
Alan, I made few minor comments in the code. Most of them could really be skipped, but one or two may be worth taking care of or at least double checked if not changed.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSCore.py:
```diff
@@ -36,16 +35,12 @@ def __init__(self, msConfig, logger=None):
-            self.phedex = PhEDEx(httpDict={'cacheduration': 0.5},
-                                 dbsUrl=dbsUrl, logger=self.logger)
+            # hard code it to production DBS otherwise PhEDEx subscribe API fails to match TMDB data
+            dbsUrl = "https://cmsweb.cern.ch/dbs/prod/global/DBSReader"
```
I think I have tested it in the past with testbed DBSReader and it was giving me some failures (likely the reason for this short comment above).
It could go in the config as well, but given that it won't change between different instances of the service, I decided simply to hard-wire it here.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -618,60 +618,30 @@ def makeTransferRequest(self, wflow):
+                                                                  subLevel, dataSize, nodes, idx)
+                if not success:
+                    # stop any other data placement for this workflow
+                    break
```
I was going to say it's already logged from the makeTransferRucio/Phedex methods, but it's a good idea to log it here as well.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -680,45 +650,143 @@ def makeTransferRequest(self, wflow):
+        :param nodeIdx: index of the node/RSE to be used in the replication rule
+        :return: a boolean flagging whether it succeeded or not, and the rule id
+        """
+        subLevel = "ALL" if subLevel == "dataset" else "DATASET"
```
`subLevel` is a subscription level in PhEDEx, which can be either `block` or at `dataset` level.
In Rucio, this is called as `grouping` level, which also mandates how data is transferred/grouped in the same destination. I'll add a comment here too.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -680,45 +650,143 @@ def makeTransferRequest(self, wflow):
+                rule['ask_approval'] = True
+
+            # Then make the data subscription, for real!!!
+            res = self.rucio.createReplicationRule(dids, rseExpr, **rule)
```
It's probably a better idea indeed. Especially because I am not sure whether this AccessDenied is a general issue related to the account/credentials we are using (this is what I assumed in the first place), or whether it could be related to RSE's that we cannot write to (or approve things)... I'll update it too

amaltaro commented:
Todor, thanks for your review. Following your review, I have made further changes in my last commit. Please have another look, and if you agree, mark those conversations as resolved.

todor-ivanov commented:
Thanks @amaltaro I took a quick look at your new commit and I think you addressed everything I mentioned. I think the changes are good to go now.

amaltaro committed:
unit tests for PycurlRucio

cleanup init.py module

more fixes to unit tests

unit test for blocks/sizes per container

fix unit tests

amaltaro committed:
default rucio rules to asynchronous

amaltaro commented:
@nsmith- and @ericvaandering I think I have finally converged on this PR and tested the most common use cases.
I'm not asking for your review because it became very big, but feel free to go through the initial description at the very top and ask/suggest anything else that you might consider important. 

I might merge it today, but feel free to leave questions/comments at any time and I can follow up on those in a different PR, if needed.

amaltaro commented:
All right, it should be tested enough by now...
